drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,319 @@
+		"access":      rule.Access,
+		"direction":   rule.Direction,
+		"protocol":    rule.Protocol,
+		"portRanges":  getDestinationPorts(rule),
+	}
+	priority := to.Int32(rule.Priority)
+	if priority > dockerPriority {
+		return nil
+	}
+	purpose := strings.ToLower(fmt.Sprintf("%s %s %s", rule.Access, rule.Direction, rule.Protocol))
+	if purpose == "deny inbound tcp" || purpose == "deny inbound *" {
+		// check if docker port is covered by rule
+		for _, pr := range getDestinationPorts(rule) {
+			portWithinRange, err := checkIfPortInPortRange(dockerPort, pr)
+			if err != nil {
+				return fmt.Errorf("%s: %s", err, f)
+			}
+			if portWithinRange {
+				return fmt.Errorf("rule %s set to %s requests to the Docker port %d: %s",
+					to.String(rule.Name), purpose, dockerPort, f)
+			}
+		}
+	}
+	return nil
+}
+
+func checkIfPortInPortRange(port int, portRange string) (bool, error) {
+	if portRange == "*" {
+		return true, nil
+	}
+	s := strings.Split(portRange, "-")
+	if len(s) == 2 {
+		// Port range represents is a range i.e. "2000-3000"
+		portMin, portMinErr := strconv.Atoi(s[0])
+		portMax, portMaxErr := strconv.Atoi(s[1])
+		if portMinErr == nil || portMaxErr == nil || portMin < portMax {
+			return (port >= portMin) && (port <= portMax), nil
+		}
+	}
+	if len(s) == 1 {
+		// Port range represents a single port i.e. "2376"
+		portVal, err := strconv.Atoi(s[0])
+		if err == nil {
+			return port == portVal, nil
+		}
+	}
+	return false, fmt.Errorf("invalid port range %s", portRange)
+}
+
+func getDestinationPorts(rule network.SecurityRule) []string {
+	ports := []string{}
+	if rule.DestinationPortRange != nil {
+		ports = append(ports, *rule.DestinationPortRange)
+	}
+	if rule.DestinationPortRanges != nil {
+		for _, port := range *rule.DestinationPortRanges {
+			ports = append(ports, port)
+		}
+	}
+	return ports
+}
+
+func removeSecurityRule(rules *[]network.SecurityRule, rule network.SecurityRule) *[]network.SecurityRule {
+	newRules := *rules
+	for i, elem := range newRules {
+		if elem != rule {
+			continue
+		}
+		if len(newRules) == 1 {
+			return &[]network.SecurityRule{}
+		}
+		lastElemIndex := len(newRules) - 1
+		newRules[i] = newRules[lastElemIndex]
+		newRules = newRules[:lastElemIndex]
+	}
+	return &newRules
+}
+
+// OpenPort is a port to be opened within an NSG
+type OpenPort struct {
+	Name  string
+	Port  string
+	Proto network.SecurityRuleProtocol
+}
+
+const (
+	machineOpenPortFormat = "Port%s-%sAllowAny"
+	// NOTE(aiyengar2): This regex is used to identify if the given rule was set by Rancher. It parses the
+	// machineOpenPortFormat that is used by Rancher to set open ports requested by the user, where the %s values are set to:
+	// 1) port: a 1-5 digit number (i.e. \d{1,5})
+	// 2) proto: a 3 letter protocol (i.e. [a-zA-Z]{1,3}) or 'Asterisk'.
+	machineOpenPortRegex = `Port(\d{1,5})-([a-zA-Z]{1,3}|Asterisk)AllowAny`
+)
+
+func isOpenPort(rule network.SecurityRule) bool {
+	re := regexp.MustCompile(machineOpenPortRegex)
```

Comment:
should move the regex to a package level variable, no need to recompile every call

Reply:
will move in the next commit!